### DISCUSSION TITLE
list_budget_rules return 200OK, but when run flask to create new API endpoint it shows: "TypeError: list_budget_rules() missing 2 required positional arguments: 'nextToken' and 'pageSize'"

### DISCUSSION THREAD

**Discussion by giangdaughtry** [denisneuf/python-amazon-ad-api#197] [General]
Hello everyone,
I already made a successful list_budget_rules request following your template. And get 200OK with payload:
// now in my amazon ad account dont have any budget rules.
 'payload': {'budgetRulesForAdvertiserResponse': [], 'nextToken': None}} 
 
 But when I run a flask to shows the result in payload, it get the error: 

 Traceback (most recent call last):
  File "D:\BUSINESS\MMO\AMZ\FBA\AmzAds\AMS Project\python-amazon-ad-api\.venv\Lib\site-packages\flask\app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\BUSINESS\MMO\AMZ\FBA\AmzAds\AMS Project\python-amazon-ad-api\.venv\Lib\site-packages\flask\app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\BUSINESS\MMO\AMZ\FBA\AmzAds\AMS Project\python-amazon-ad-api\.venv\Lib\site-packages\flask\app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\BUSINESS\MMO\AMZ\FBA\AmzAds\AMS Project\python-amazon-ad-api\.venv\Lib\site-packages\flask\app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: list_budget_rules() missing 2 required positional arguments: 'nextToken' and 'pageSize'
2024-04-27 05:59:21,571:INFO:127.0.0.1 - - [27/Apr/2024 05:59:21] "GET /sp/list_budget_rules HTTP/1.1" 500 -
2024-04-27 05:59:21,723:ERROR:Exception on /sp/list_budget_rules [GET]
Traceback (most recent call last):
  File "D:\BUSINESS\MMO\AMZ\FBA\AmzAds\AMS Project\python-amazon-ad-api\.venv\Lib\site-packages\flask\app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\BUSINESS\MMO\AMZ\FBA\AmzAds\AMS Project\python-amazon-ad-api\.venv\Lib\site-packages\flask\app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\BUSINESS\MMO\AMZ\FBA\AmzAds\AMS Project\python-amazon-ad-api\.venv\Lib\site-packages\flask\app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\BUSINESS\MMO\AMZ\FBA\AmzAds\AMS Project\python-amazon-ad-api\.venv\Lib\site-packages\flask\app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: list_budget_rules() missing 2 required positional arguments: 'nextToken' and 'pageSize

**Comment by denisneuf** [comment]
Could you please share your function to call the endpoint?

some endpoints in general return a subset of data based on the pagesize parameter, e.g. there are enough data as 200 results but the maximum pagesize is 10, the first request returns a nextToken which is necessary to use to complete the data collection.

You can set your own pageSize or the endpoint will use the default of the api endpoint, but if the total result data is over the pageSize it will return in the payload nextToken that you will need to use in the next request to get the following subset, eg: first call returns for 1 to 10, second, from 11 -20,....and so on

This is very easy to deal with a decorator like in the Utils:

@Utils.load_all_pages(throttle_by_seconds=1, next_token_param="nextToken")
def get_list_invoices(**kwargs):
    return Invoices(account=store, marketplace=marketplace, debug=True).list_invoices(**kwargs)

**Reply by giangdaughtry** [reply]
I got it. Thank you. Now, Im facing the issue for the CORS. I setup a proxy at localhost:5000 for localhost:3000
Now i want to add three into header of response from server.
('Access-Control-Allow-Origin', '*')
('Access-Control-Allow-Headers', '*')
('Access-Control-Allow-Methods', '*')

How can I do this? 
For example: GET method of:
@sp_endpoint('/sp/budgetRules', method='GET')
    def list_budget_rules(self, **kwargs) -> ApiResponse:

**Comment by denisneuf** [comment]
The client supports proxy:

proxies: Optional[Dict[str, str]] = None,

But in last term it delegates in the library request:

from requests import request

r = requests.get(
    location,
    headers=headers or self.headers,
    data=None,
    allow_redirects=True,
    timeout=self.timeout,
    proxies=self.proxies,
    verify=self.verify,
)

So it will pass whatever you set  when creating your client:


my_proxies = {
    'http': 'http://10.10.1.10:3128',
    'https': 'https://10.10.1.11:1080',
}

try:

    result = sponsored_products.Campaigns(account=store, marketplace=marketplace, proxies=my_proxies, debug=True).list_campaigns_extended(
    )

    campaigns = result.payload

except AdvertisingApiException as error:
    logging.info(error)
Kindest regards,

Daniel

**Reply by giangdaughtry** [reply]
Thanks a lot.

**Comment by giangdaughtry** [comment]
Now I am creating REACT front end to handle POST fetch API to create new budget rule. I am thinking of using Flask. But I dont know how to transfer values from inputs front end to make a POST fetch. 
def create_budget_rules(self, **kwargs) -> ApiResponse:
        r"""
        Creates one or more budget rules.

        Request Body
            | CreateSPBudgetRulesRequest {
            | **budgetRulesDetails** (array) [
            |   maxItems: 25.
            |   A list of budget rule details.
            |   SPBudgetRuleDetails {
            |   Object representing details of a budget rule for SP campaign
            |   **duration** RuleDuration {
            |       eventTypeRuleDuration EventTypeRuleDuration {
            |           Object representing event type rule duration.
            |           eventId*(string): The event identifier. This value is available from the budget rules recommendation API.
            |           endDate(string): The event end date in YYYYMMDD format. Read-only.
            |           eventName(string): The event name. Read-only.
            |           startDate(string): The event start date in YYYYMMDD format. Read-only. Note that this field is present only for announced events.
            |           }
            |       dateRangeTypeRuleDuration DateRangeTypeRuleDuration {
            |           Object representing date range type rule duration.
            |           endDate(string): The end date of the budget rule in YYYYMMDD format. The end date is inclusive. Required to be equal or greater than `startDate`.
            |           startDate*(string): The start date of the budget rule in YYYYMMDD format. The start date is inclusive. Required to be greater than or equal to current date.
            |           eventName(string): The event name. Read-only.
            |           startDate(string): The event start date in YYYYMMDD format. Read-only. Note that this field is present only for announced events.
            |           }
            |       }
            | **recurrence** Recurrence {
            |       type (string): The frequency of the rule application. Enum: ['DAILY']
            |       daysOfWeek (array). Object representing days of the week for weekly type rule. It is not required for daily recurrence type
            |       DayOfWeek [
            |           DayOfWeek(string): The day of the week. Enum: ['MONDAY', 'TUESDAY', 'WEDNESDAY', 'THURSDAY', 'FRIDAY', 'SATURDAY', 'SUNDAY']
            |           ]
            |       }
            | **ruleType**\* SPRuleType (string): The type of budget rule. SCHEDULE: A budget rule based on a start and end date. PERFORMANCE: A budget rule based on advertising performance criteria. Enum: ['SCHEDULE', 'PERFORMANCE']
            | **budgetIncreaseBy** budgetIncreaseBy {
            |       type* BudgetChangeType (string): The value by which to update the budget of the budget rule. Enum: ['PERCENT']
            |       value* number($double): The budget value.
            |       }
            | **name** (string): The budget rule name. Required to be unique within a campaign. maxLength: 355
            | **performanceMeasureCondition**\* PerformanceMeasureCondition {
            |       metricName* PerformanceMetrics (string): The advertising performance metric. Enum: [ ACOS, CTR, CVR, ROAS ]
            |       comparisonOperator* ComparisonOperator (string): The comparison operator. Enum: [ GREATER_THAN, LESS_THAN, EQUAL_TO, LESS_THAN_OR_EQUAL_TO, GREATER_THAN_OR_EQUAL_TO ]
            |       threshold* number($double): The performance threshold value.
            |       }
            |    ]
            | }


        Returns:

            ApiResponse
